### PR TITLE
Implemented reusing of BigQueryReadClient and BigQueryWriteClient instances

### PR DESCRIPTION
**Pull request by VinayLondhe14**

### CODE REVIEW

**Comment by emkornfield** [comment]
Can you run some benchmarks (tpc-h/tpc-ds) to see if this impacts perf (and share internally)?

**Review by davidrabinowitz — CHANGES_REQUESTED** [review]

File: bigquery-connector-common/src/main/java/com/google/cloud/bigquery/connector/common/BigQueryClientFactory.java [hunk 1]
```diff
@@ -39,7 +50,78 @@ public BigQueryClientFactory(
...
+  @Override
+  public int hashCode() {
+    // Here, credentials is an instance of GoogleCredentials which can be one out of
+    // GoogleCredentials, UserCredentials, ServiceAccountCredentials, ExternalAccountCredentials or
+    // ImpersonatedCredentials (See the class BigQueryCredentialsSupplier which supplies these
+    // Credentials). Subclasses of the abstract class ExternalAccountCredentials do not have the
+    // hashCode method defined on them and hence we only compute the hashcode using
+    // userAgentHeaderProvider and bqConfig here. In the equals implementation, we serialize and
+    // compare byte arrays if either of the credentials are instances of ExternalAccountCredentials
+    if (credentials instanceof ExternalAccountCredentials) {
```
Comment (davidrabinowitz) [inline bigquery-connector-common/src/main/java/com/google/cloud/bigquery/connector/common/BigQueryClientFactory.java:89]: `hashCode()` input should be the same as `equals()`, so please create a hashCode for `ExternalAccountCredentials` as well. Th e best is to move all credentials handling code (including the equals related methods below) to an independent utility class - none of this functionality is unique to `BigQueryClientFactory`

File: bigquery-connector-common/src/main/java/com/google/cloud/bigquery/connector/common/BigQueryClientFactory.java [hunk 2]
```diff
@@ -86,4 +168,40 @@ public void setProxyConfig(InstantiatingGrpcChannelProvider.Builder transportBui
...
+      return Objects.equal(credentials1, credentials2);
+    }
+
+    try {
+      ByteArrayOutputStream byteArrayOutputStream1 = new ByteArrayOutputStream();
+      ObjectOutputStream objectOutputStream1 = new ObjectOutputStream(byteArrayOutputStream1);
+      objectOutputStream1.writeObject(credentials1);
+      objectOutputStream1.close();
+
+      ByteArrayOutputStream byteArrayOutputStream2 = new ByteArrayOutputStream();
```
Comment (davidrabinowitz) [inline bigquery-connector-common/src/main/java/com/google/cloud/bigquery/connector/common/BigQueryClientFactory.java:196]: Duplicated code, lines 191-194, 196-199 can be extracted to another method returning `byte[]` (which can also be used by `hashCode()`, see above)

**Review by davidrabinowitz — CHANGES_REQUESTED** [review]

**Review by davidrabinowitz — APPROVED** [review]

**Pull request merged**